### PR TITLE
mon: fix `fs set down` to adjust max_mds only when cluster is not down

### PR DESCRIPTION
**Pull request by littlebees**
Applying `fs set down false` on an up cluster will set the cluster's max_mds to 1, regardless of the cluster's current max_mds.

`fs set down false` should only change max_mds when the cluster was set to down. Otherwise, the cluster should remain unchanged.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/mon/FSCommands.cc [hunk 1]
```diff
@@ -598,6 +598,11 @@ int FileSystemCommandHandler::set_val(Monitor *mon, FSMap& fsmap, MonOpRequestRe
 
       ss << fsp->get_mds_map().get_fs_name();
 
+      if (!is_down && fsp->get_mds_map().get_old_max_mds() == 0) {
```
Comment (batrick) [inline src/mon/FSCommands.cc:601]:
```suggestion
      if (!is_down && fsp->get_mds_map().get_max_mds() > 0) {
```

would work, no?

**Review by batrick — CHANGES_REQUESTED** [review]

File: src/mon/FSCommands.cc (hunk 1 above)
Comment (dparmar18) [reply src/mon/FSCommands.cc:601]: +1

**Comment by dparmar18** [comment]
This looks be to intentional, [cephfs-admin-docs](https://docs.ceph.com/en/latest/cephfs/administration/#taking-the-cluster-down)  say:
```
This will also restore the previous value of max_mds. MDS daemons are brought down in a way such that journals are flushed to the metadata pool and all client I/O is stopped.
```

If this behaviour is to be changed then the docs would also need to be updated. https://github.com/ceph/ceph/commit/4e9ffffb95156f392926abc63bbff637d3bcba70 brought in this documentation.

**Comment by dparmar18** [comment]
So, when a CephFS cluster is taken down using `ceph fs set <fs_name> down true`, it does this:
https://github.com/ceph/ceph/blob/5969c1192fb5679fd2d897d7c8e208ea7f1800ff/src/mon/FSCommands.cc#L604-L608

It first records the old `max_mds` val using:
```
void set_old_max_mds() { old_max_mds = max_mds; }
```
and then `max_mds` is set to `0`. Which means first the `max_mds` is preserved and then reset to `0`. Looks correct.

So, now when `ceph fs set <fs_name> down false` is called:
https://github.com/ceph/ceph/blob/5969c1192fb5679fd2d897d7c8e208ea7f1800ff/src/mon/FSCommands.cc#L609-L612

old `max_mds` val is fetched; if it is `0` i.e. the old `max_mds` was `0` (huh?), it sets the new max_mds to `1`, maybe it could be possible that the `max_mds` was always `1` **OR** some cmd was ran between `ceph fs set <fs_name> down true` and `ceph fs set <fs_name> down false` which reset the `max_mds` to `1`?

It looks to me that this is correct functionality is correct, it doesn't always set the `max_mds` to `1` but follows what the last recorded value is and sets it.

**Comment by dparmar18** [comment]
> So, when a CephFS cluster is taken down using `ceph fs set <fs_name> down true`, it does this:
...
...
...
> It looks to me that this functionality is correct, it doesn't always set the `max_mds` to `1` but follows what the last recorded value is and sets it.

@batrick any thoughts?

**Review by rishabh-d-dave — CHANGES_REQUESTED** [review]
I couldn't reproduce the bug this PR is trying to fix, it doesn't exist.  `ceph fs set a down true` successfully brought down the 2 active MDSs FS had and `ceph fs set a down false` successfully brought FS up with same number of active MDSs.

EDIT
@littlebees Was this bug seen in one of the release branches?

EDIT 2
**The bug is reproducible** on a new cluster when `ceph fs set a down false` is run without running `ceph fs set a down true` in the past.

**Comment by rishabh-d-dave** [comment]
> It looks to me that this functionality is correct, it doesn't always set the max_mds to 1 but follows what the last recorded value is and sets it.

Ugh! Seeing this Dhairya's comment now since the page was refreshed while posting my review.

Agreed +1.

@dparmar18 @batrick

**Comment by vshankar** [comment]
> This looks be to intentional, [cephfs-admin-docs](https://docs.ceph.com/en/latest/cephfs/administration/#taking-the-cluster-down) say:
> 
> ```
> This will also restore the previous value of max_mds. MDS daemons are brought down in a way such that journals are flushed to the metadata pool and all client I/O is stopped.
> ```
> 
> If this behaviour is to be changed then the docs would also need to be updated. [4e9ffff](https://github.com/ceph/ceph/commit/4e9ffffb95156f392926abc63bbff637d3bcba70) brought in this documentation.

I think the bug that this change is fixing is when `fs set <> down false` when the file system is online is causing max_mds to get reset to one (1).

**Comment by dparmar18** [comment]
> > This looks be to intentional, [cephfs-admin-docs](https://docs.ceph.com/en/latest/cephfs/administration/#taking-the-cluster-down) say:
> > ```
> > This will also restore the previous value of max_mds. MDS daemons are brought down in a way such that journals are flushed to the metadata pool and all client I/O is stopped.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > If this behaviour is to be changed then the docs would also need to be updated. [4e9ffff](https://github.com/ceph/ceph/commit/4e9ffffb95156f392926abc63bbff637d3bcba70) brought in this documentation.
> 
> I think the bug that this change is fixing is when `fs set <> down false` when the file system is online is causing max_mds to get reset to one (1).

The code never resets to one unless the last recorded max_mds is `0`. check https://github.com/ceph/ceph/pull/58582#issuecomment-2248638255

**Comment by dparmar18** [comment]
> > This looks be to intentional, [cephfs-admin-docs](https://docs.ceph.com/en/latest/cephfs/administration/#taking-the-cluster-down) say:
> > ```
> > This will also restore the previous value of max_mds. MDS daemons are brought down in a way such that journals are flushed to the metadata pool and all client I/O is stopped.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > If this behaviour is to be changed then the docs would also need to be updated. [4e9ffff](https://github.com/ceph/ceph/commit/4e9ffffb95156f392926abc63bbff637d3bcba70) brought in this documentation.
> 
> I think the bug that this change is fixing is when `fs set <> down false` when the file system is online is causing max_mds to get reset to one (1).

My bad, I understand this now. You're right.

**Comment by dparmar18** [comment]
this can be easily reproduced, better have a test case included as well.

File: src/mon/FSCommands.cc [hunk 2]
```diff
@@ -598,6 +598,11 @@ int FileSystemCommandHandler::set_val(Monitor *mon, FSMap& fsmap, MonOpRequestRe
 
       ss << fsp->get_mds_map().get_fs_name();
 
+      if (!is_down && fsp->get_mds_map().get_max_mds() > 0) {
+        ss << " already up";
```
Comment (dparmar18) [inline src/mon/FSCommands.cc:602]: This statement can be bettered, something like:

```
ss << " is already online, trying to bring up an already running CepFS cluster has no impact, run `ceph fs set " << fsp->get_mds_map().get_fs_name() << " down true` first."; 
```

**Comment by rishabh-d-dave** [comment]
jenkins test api

**Comment by rishabh-d-dave** [comment]
jenkins test make check

**Review by batrick — APPROVED** [review]

File: src/mon/FSCommands.cc [hunk 3]
```diff
@@ -598,6 +598,11 @@ int FileSystemCommandHandler::set_val(Monitor *mon, FSMap& fsmap, MonOpRequestRe
 
       ss << fsp->get_mds_map().get_fs_name();
 
+      if (!is_down && fsp->get_mds_map().get_max_mds() > 0) {
+        ss << " is already online, trying to bring up an already running CepFS cluster has no impact, run `ceph fs set " << fsp->get_mds_map().get_fs_name() << " down true` first.";
```
Comment (batrick) [inline src/mon/FSCommands.cc:602]:
```suggestion
        ss << " is already online";
```

**Review by batrick — CHANGES_REQUESTED** [review]
Sorry, I disagree with the error message change. We shouldn't suggest the operator bring the fs down just so that this command will succeed. That doesn't make sense.

**Review by dparmar18 — APPROVED** [review]

**Review by batrick — APPROVED** [review]

**Comment by batrick** [comment]
This PR is under test in [https://tracker.ceph.com/issues/67214](https://tracker.ceph.com/issues/67214).

**Pull request merged**